session.go

```diff
@@ -284,8 +286,22 @@ func (s *Session) AcceptStream() (*Stream, error) {
 }
 
 // Close is used to close the session and all streams.
-// Attempts to send a GoAway before closing the connection.
+// Attempts to send a GoAway before closing the connection. The GoAway may not actually be sent depending on the
+// semantics of the underlying net.Conn. For TCP connections, it may be dropped depending on LINGER value or
+// if there's unread data in the kernel receive buffer.
 func (s *Session) Close() error {
+	return s.close(ErrSessionShutdown, true, goAwayNormal)
+}
+
+// CloseWithError is used to close the session and all streams after sending a GoAway message with errCode.
+// The GoAway may not actually be sent depending on the semantics of the underlying net.Conn.
+// For TCP connections, it may be dropped depending on LINGER value or if there's unread data in the kernel
+// receive buffer.
+func (s *Session) CloseWithError(errCode uint32) error {
```

Comment:
Have we updated the connection manager to be able to deal with potential blocking here? Also, we should probably document it.